sdk/tests/conformance2/textures/misc/copy-texture-image-webgl-specific.html

```diff
@@ -174,6 +174,140 @@
+        var bytes = new Uint8Array(width2 * height2 * 4);
+        gl.readPixels(0, 0, width2, height2, gl.RGBA, gl.UNSIGNED_BYTE, bytes);
+        wtu.glErrorShouldBe(gl, gl.NO_ERROR, "readpixel should succeed.");
+
+        var sourceX = new Object();
+        var sourceY = new Object();
+        Clip(xx, copyWidth, width, sourceX);
+        Clip(yy, copyHeight, height, sourceY);
+        var destX = sourceX.start - xx + xoffset;
+        var rangeX = sourceX.range;
+        var destY = sourceY.start - yy + yoffset;
+        var rangeY = sourceY.range;
+        for (var y = 0; y < height2; y++) {
+            for (var x = 0; x < width2 * 4; x++) {
+                // pixels copied from read framebuffer should be 0x01
+                if (x >= destX * 4 && x < (destX + rangeX) * 4 && y >= destY && y < destY + rangeY) {
+                    if (bytes[y * height2 * 4 + x] != 0x01) {
+                        testFailed("byte comparison failure, byte at "+ (y * height2 * 4 + x) + " was " +
+                        bytes[y * height2 * 4 + x] +", should be 1");
+                        break;
+                     }
+                // pixels out-of-bounds should untouched
+                } else {
+                    if (bytes[y * height2 * 4 + x] != 0xFF) {
+                        testFailed("byte comparison failure, byte at "+ (y * height2 * 4 + x) + " was " +
+                        bytes[y * height2 * 4 + x] + ", should be 255");
+                        break;
+                    }
+                }
+            }
+            if (x < width2 * 4) {
+                break;
+            }
+        }
+    } else {
+        testFailed("framebuffer not complete");
+    }
+
+    gl.bindTexture(gl.TEXTURE_2D, null);
+    gl.bindTexture(gl.TEXTURE_3D, null);
+    gl.bindFramebuffer(gl.FRAMEBUFFER, null);
+    gl.deleteFramebuffer(fbo);
+    gl.deleteTexture(texture[0]);
+    gl.deleteTexture(texture[1]);
+}
+
+function copytexsubimage3d_out_of_bounds() {
+    debug("");
+    debug("Test pixels outside of read framebuffer for CopyTexSubImage3D");
+
+    for(var i=0; i < testlist.length; i++) {
+        copytexsubimage3d_test(testlist[i][0], testlist[i][1], testlist[i][2], testlist[i][3]);
+    }
+};
+
+/**
+ * This array defines some copy areas for CopyTexSubImage3D.
+ * A copy area is defined by x coordinate, y coordinate, copyWidth and copyHeight.
+ * five kinds of copy areas are listed as following:
+ * 1. all range of read framebuffer is totally covered by copy area.
+ * 2. one angle and two edges of copy area are overlapped with those of read framebuffer, while other directions are outside of read framebuffer.
+ * 3. two angles and three edges of copy area are overlapped with those of read framebuffer, while other directions are outside of read framebuffer.
+ * 4. copy area intersects with some part of read framebuffer, but no corners and edges are overlapped.
+ * 5. all pixels are outside of read framebuffer. No pixel is copied at all.
```

Comment:
This comment left me more confused than without.

Maybe simply say the source framebuffer is of size (2, 2)? No need to enumerate all five cases as the data is self-apparent.